src/Compilers/Core/Portable/SourceGeneration/GeneratorState.cs

```diff
@@ -2,31 +2,103 @@
 
+using System;
 using System.Collections.Immutable;
+using System.Diagnostics;
 
 #nullable enable
 namespace Microsoft.CodeAnalysis
 {
+    /// <summary>
+    /// Represents the current state of a generator
+    /// </summary>
+    /// <remarks>
+    /// A generator state is essentially a small state machine:
+    ///     1. We start with just info
+    ///     2. We can optionally set a receiver that will be used in the next iteration
+    ///     3. We set either the result or error of the iteration that just happened
+    ///     4. We either go back to state 2 or 3.
+    /// </remarks>
     internal readonly struct GeneratorState
     {
         public GeneratorState(GeneratorInfo info)
-            : this(info, ImmutableDictionary<GeneratedSourceText, SyntaxTree>.Empty)
+            : this(info, ImmutableArray<GeneratedSourceText>.Empty, ImmutableArray<SyntaxTree>.Empty, ImmutableArray<Diagnostic>.Empty, syntaxReceiver: null, exception: null)
         {
         }
 
-        private GeneratorState(GeneratorInfo info, ImmutableDictionary<GeneratedSourceText, SyntaxTree> sources)
+        private GeneratorState(GeneratorInfo info, ImmutableArray<GeneratedSourceText> sourceTexts, ImmutableArray<SyntaxTree> trees, ImmutableArray<Diagnostic> diagnostics, ISyntaxReceiver? syntaxReceiver, Exception? exception)
         {
-            this.Sources = sources;
+            this.SourceTexts = sourceTexts;
+            this.Trees = trees;
             this.Info = info;
+            this.Diagnostics = diagnostics;
+            this.SyntaxReceiver = syntaxReceiver;
+            this.Exception = exception;
         }
 
-        internal ImmutableDictionary<GeneratedSourceText, SyntaxTree> Sources { get; }
+        internal ImmutableArray<GeneratedSourceText> SourceTexts { get; }
+
+        internal ImmutableArray<SyntaxTree> Trees { get; }
 
         internal GeneratorInfo Info { get; }
 
-        internal GeneratorState WithSources(ImmutableDictionary<GeneratedSourceText, SyntaxTree> sources)
+        internal ISyntaxReceiver? SyntaxReceiver { get; }
+
+        internal Exception? Exception { get; }
+
+        internal ImmutableArray<Diagnostic> Diagnostics { get; }
+
+        /// <summary>
+        /// Called to save the created syntax receiver
+        /// </summary>
+        /// <remarks>
+        /// We retain the existing state, as it may be needed for the upcoming generation pass.
+        /// </remarks>
+        internal GeneratorState WithReceiver(ISyntaxReceiver syntaxReceiver)
+        {
+            Debug.Assert(this.SyntaxReceiver is null);
+            return new GeneratorState(this.Info,
+                                      sourceTexts: this.SourceTexts,
+                                      trees: this.Trees,
+                                      diagnostics: this.Diagnostics,
+                                      syntaxReceiver: syntaxReceiver,
+                                      exception: null);
```

Comment:
Why doesn't this clear the results here? This effectively moves us back to state 2 hence expectation is that results should be cleared.

Reply:
Because if you use partial generation (admittedly not exposed) then you want to have access to the results from the previous run. You're just going to run with a new syntax receiver. I'm begining to think this should be two structs, one for  'state before generation' and one for 'state post generation', but I'm not convinced its worth breaking apart right now (especially as its internal).